### PR TITLE
VisualShader: Add LinearToSRGB and SRGBToLinear to ColorFunc node

### LINKED ISSUE #10818: VisualShader: Add LinearToSRGB and SRGBToLinear to the ColorFunc node
### Describe the project you are working on

Writing shaders

### Describe the problem or limitation you are having in your project

*Usually* you will never need to convert between linear and SRGB color spaces in user shaders. However, there are a lot of steps along the pipeline and sometimes its useful to use these to check your work. There are also cases where you *do* want want convert to and from SRGB. Occasionally, SRGB/linear bugs do happen in the actual engine, too, and these nodes can be used to diagnose.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add conversion functions, LinearToSRGB and SRGBToLinear, to the ColorFunc visual shader node.

See also https://github.com/godotengine/godot-proposals/issues/9150.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Copy the linear/srgb functions used by godot, and add them to the ColorFunc node in visualshader.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You could use an expression node with the conversion function. Or, the simple approximations `pow(x, 2.2)` and `pow(x, .45)` do work well enough for simple diagnosis of color space mismatches.

### Is there a reason why this should be core and not an add-on in the asset library?

Extends the functionality of an existing node.

### PR DESCRIPTION
**Pull request by tetrapod00**
Implements and closes https://github.com/godotengine/godot-proposals/issues/10818.

Adds LinearToSRGB and SRGBToLinear conversions to the visual shader node `ColorFunc`.

godot windows editor dev x86_64_5xYhLe22OV

**Implementation:**
This node automatically switches it's conversion function depending on which renderer is used.

The RenderingDevice (Forward+ and Mobile Renderer) conversion functions:
https://github.com/godotengine/godot/blob/4254946de93bab0cc1fb36a7b9039c9ec43be924/servers/rendering/renderer_rd/shaders/effects/copy_to_fb.glsl#L121-L130

The Compatibility renderer uses slightly different functions:
https://github.com/godotengine/godot/blob/4254946de93bab0cc1fb36a7b9039c9ec43be924/drivers/gles3/shaders/tonemap_inc.glsl#L13-L26

If https://github.com/godotengine/godot/pull/90187 is merged, *and* the change in that PR is also made to the GLSL version of the `linear_to_srgb` and `srgb_to_linear` functions, then this visual shader node will need to be updated.

### CODE REVIEW

File: scene/resources/visual_shader_nodes.cpp [hunk 1]
```diff
@@ -3273,12 +3286,14 @@ void VisualShaderNodeColorFunc::_bind_methods() {
...
 	ClassDB::bind_method(D_METHOD("get_function"), &VisualShaderNodeColorFunc::get_function);
 
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "function", PROPERTY_HINT_ENUM, "Grayscale,HSV2RGB,RGB2HSV,Sepia"), "set_function", "get_function");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "function", PROPERTY_HINT_ENUM, "Grayscale,HSV2RGB,RGB2HSV,Sepia,LinearToSRGB,SRGBToLinear"), "set_function", "get_function");
 
 	BIND_ENUM_CONSTANT(FUNC_GRAYSCALE);
 	BIND_ENUM_CONSTANT(FUNC_HSV2RGB);
 	BIND_ENUM_CONSTANT(FUNC_RGB2HSV);
 	BIND_ENUM_CONSTANT(FUNC_SEPIA);
+	BIND_ENUM_CONSTANT(FUNC_LINEAR_TO_SRGB);
```
Comment (tetrapod00) [inline scene/resources/visual_shader_nodes.cpp:3295]: These enums could be `FUNC_LINEAR2SRGB` and displayed as "Linear2SRGB", to match `FUNC_HSV2RGB` and "HSV2RGB". I think spelling out "To" is more legible in this case but I understand if consistency is more important.

Comment (JoNax97) [reply scene/resources/visual_shader_nodes.cpp:3295]: I think there's been a push to move away from using 2 since Godot 4

**Review by Geometror — COMMENTED** [review]
Originally, I wanted to suggest switching the computation based on the current renderer, but after looking at the referenced blog article, I think it would be fine to just use the compatibility version (since it's a very good approximation). Maybe it can be used for RD renderers as well. CC @clayjohn

**Comment by tetrapod00** [comment]
All else being equal, the visual shader node should be accurate to whatever computation is used in the current renderer.
But if the Compatibility approximation is both much faster and still very accurate, then it makes sense to use it for visual shdaer nodes, where you might be converting between color spaces repeatedly (if a node is there, it will be used!).

**Comment by tetrapod00** [comment]
The computation now depends on the renderer, exactly matching whichever computation is used internally.

**Review by QbieShay — APPROVED** [review]

**Review by Geometror — APPROVED** [review]

File: editor/plugins/visual_shader_editor_plugin.cpp [hunk 2]
```diff
@@ -6777,6 +6777,8 @@ VisualShaderEditor::VisualShaderEditor() {
 	add_options.push_back(AddOption("HSV2RGB", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts HSV vector to RGB equivalent."), { VisualShaderNodeColorFunc::FUNC_HSV2RGB, VisualShaderNodeVectorFunc::OP_TYPE_VECTOR_3D }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
 	add_options.push_back(AddOption("RGB2HSV", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts RGB vector to HSV equivalent."), { VisualShaderNodeColorFunc::FUNC_RGB2HSV, VisualShaderNodeVectorFunc::OP_TYPE_VECTOR_3D }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
 	add_options.push_back(AddOption("Sepia", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Sepia function."), { VisualShaderNodeColorFunc::FUNC_SEPIA }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
+	add_options.push_back(AddOption("LinearToSRGB", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts color from linear to sRGB color space."), { VisualShaderNodeColorFunc::FUNC_LINEAR_TO_SRGB }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
```
Comment (Chaosus) [inline editor/plugins/visual_shader_editor_plugin.cpp:6780]:
```suggestion
	add_options.push_back(AddOption("HSV2RGB", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts HSV vector to RGB equivalent."), { VisualShaderNodeColorFunc::FUNC_HSV2RGB, VisualShaderNodeVectorFunc::OP_TYPE_VECTOR_3D }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
	add_options.push_back(AddOption("LinearToSRGB", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts color from linear to sRGB color space."), { VisualShaderNodeColorFunc::FUNC_LINEAR_TO_SRGB }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
	add_options.push_back(AddOption("RGB2HSV", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Converts RGB vector to HSV equivalent."), { VisualShaderNodeColorFunc::FUNC_RGB2HSV, VisualShaderNodeVectorFunc::OP_TYPE_VECTOR_3D }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
	add_options.push_back(AddOption("Sepia", "Color/Functions", "VisualShaderNodeColorFunc", TTR("Sepia function."), { VisualShaderNodeColorFunc::FUNC_SEPIA }, VisualShaderNode::PORT_TYPE_VECTOR_3D));
```

Entries must follow the alphabet order.

**Review by Mickeon — APPROVED** [review]
Assuming accuracy, the docs are completely fine.

File: scene/resources/visual_shader_nodes.cpp (hunk 1 above)
Comment (tetrapod00) [reply scene/resources/visual_shader_nodes.cpp:3305]: Seems like consensus is that "to" is better than "2".

**Comment by Repiteo** [comment]
Thanks!

**Comment by allenwp** [comment]
I have been working on the [HDR output](https://github.com/godotengine/godot/pull/94496) PR, which requires the use of HDR buffers. This PR is useful because it allows simulation of traditional blending with nonlinear sRGB-encoded values, while still using HDR buffers. This approach involves performing a linear to sRGB conversion, applying blending or other shader operations, and then converting back to linear.

Especially when using HDR output it is important that all HDR values are maintained (values above `1.0`). This PR clamps these values so that colour information is lost if it is above `1.0`, but only in the Mobile and Forward+ rendering methods that most need this colour information.

I understand that "this is the way it was previously done in many shader files in Godot", but in these previous cases it wouldn't have an impact because the colour value was immediately written to a buffer that clamps to [0, 1]. The use case for this node is different: it is expected that the user may be converting to and from nonlinear-sRGB encoding to achieve different effects. If they want to clamp, they can do so with the `Clamp` visual node.

Is there a reason that the clamp function was included in this node for Mobile and Forward+?

**Comment by allenwp** [comment]
> Is there a reason that the clamp function was included in this node for Mobile and Forward+?

Since I haven't heard back after a few weeks, I've opened an [issue](https://github.com/godotengine/godot/issues/113952) and [PR](https://github.com/godotengine/godot/pull/113956) to remove clamping from the LinearToSRGB visual shader node.

**Pull request merged**